validation/src/header.rs

```diff
@@ -607,4 +620,72 @@ mod test {
             Err(ValidateHeaderError::TargetDifficultyAboveMax)
         ));
     }
+
+    fn genesis_header(bits: u32) -> BlockHeader {
+        BlockHeader {
+            version: 1,
+            prev_blockhash: Default::default(),
+            merkle_root: Default::default(),
+            time: 1296688602,
+            bits,
+            nonce: 0,
+        }
+    }
+
+    fn next_block_header(prev: BlockHeader, bits: u32) -> BlockHeader {
+        BlockHeader {
+            prev_blockhash: prev.block_hash(),
+            time: prev.time + TEN_MINUTES,
+            bits,
+            ..prev
+        }
+    }
+
+    /// Creates a chain of headers with the given length and
+    /// proof of work for the first header.
+    fn create_chain(
+        network: &Network,
+        initial_pow: u32,
+        chain_length: u32,
+    ) -> (SimpleHeaderStore, BlockHeader) {
+        let pow_limit = pow_limit_bits(network);
+        let h0 = genesis_header(initial_pow);
+        let mut store = SimpleHeaderStore::new(h0, 0);
+        let mut last_header = h0;
+
+        for _ in 1..chain_length {
+            let new_header = next_block_header(last_header, pow_limit);
+            store.add(new_header);
+            last_header = new_header;
+        }
+
+        (store, last_header)
+    }
+
+    const SOME_NON_LIMIT_POW: u32 = 7;
+    const CHAIN_LENGTH_MAX: u32 = 10;
+
+    #[test]
+    fn test_find_next_difficulty_in_chain_pow_found() {
+        // This test checks the chain of headers of different lengths
+        // with non-limit PoW in the first block header and PoW limit
+        // in all the other headers.
+        // Expect difficulty to be equal to the non-limit PoW.
+
+        // Arrange.
+        let network = Network::Regtest;
+        let expected_pow = SOME_NON_LIMIT_POW;
+
+        for chain_length in 1..CHAIN_LENGTH_MAX {
+            let (store, last_header) = create_chain(&network, expected_pow, chain_length);
+            assert_eq!(store.height() + 1, chain_length);
+
+            // Act.
+            let difficulty =
+                find_next_difficulty_in_chain(&network, &store, &last_header, chain_length - 1);
+
+            // Assert.
+            assert_eq!(difficulty, expected_pow);
+        }
+    }
```

Comment:
I think this test should be at the level of `get_next_target` and not calling `find_next_difficulty_in_chain`. The latter is a small helper method and doesn't capture the big picture. It's also poorly named, but that's unrelated to this change.

```suggestion
    #[test]
    fn test_next_target_regtest() {
        // This test checks the chain of headers of different lengths
        // with non-limit PoW in the first block header and PoW limit
        // in all the other headers.
        // Expect difficulty to be equal to the non-limit PoW.

        // Arrange.
        let network = Network::Regtest;
        let expected_pow = SOME_NON_LIMIT_POW;

        for chain_length in 1..CHAIN_LENGTH_MAX {
            let (store, last_header) = create_chain(&network, expected_pow, chain_length);
            assert_eq!(store.height() + 1, chain_length);

            // Act.
            let target = get_next_target(
                &network,
                &store,
                &last_header,
                chain_length - 1,
                &next_block_header(last_header, expected_pow)
            );

            // Assert.
            assert_eq!(target, BlockHeader::u256_from_compact_target(expected_pow));
        }
    }

```

Reply:
Done.